### PR TITLE
Issue #7117: fix package info javadoc check

### PR DESCRIPTION
**Pull request by strkkk**
#7117

report #2 https://strkkk.github.io/checkstyle/7117_2/index.html

### CODE REVIEW

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingjavadocpackage/nojavadoc/annotation/blockcomment/package-info.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+/*
+*  Some comment
```
Comment (pbludov) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingjavadocpackage/nojavadoc/annotation/blockcomment/package-info.java]: Please add `violation` marker here. This will help to separate the positive inputs from the negative ones for those who will work with this in the future.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingjavadocpackage/nojavadoc/detached/package-info.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+/**
+ * Dangling javadoc
```
Comment (pbludov) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingjavadocpackage/nojavadoc/detached/package-info.java]: (violation)

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingjavadocpackage/nojavadoc/annotation/blockcomment/package-info.java (hunk 1 above)
Comment (strkkk) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingjavadocpackage/nojavadoc/annotation/blockcomment/package-info.java]: I put violation comment on package def line (because violation is reported on this token)

Also added violation comment for other tests for this check

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingjavadocpackage/nojavadoc/detached/package-info.java (hunk 2 above)
Comment (strkkk) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingjavadocpackage/nojavadoc/detached/package-info.java]: done, see above

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java [hunk 3]
```diff
@@ -107,24 +107,31 @@ public void visitToken(DetailAST ast) {
...
+        final boolean hasBefore = ast.getPreviousSibling() != null
+            && isJavadoc(ast.getPreviousSibling());
+        final Optional<DetailAST> firstAnnotationChild = Optional.of(ast.getFirstChild())
             .map(DetailAST::getFirstChild)
-            .map(DetailAST::getFirstChild)
-            .map(MissingJavadocPackageCheck::isJavadocComment)
-            .orElse(false);
+            .map(DetailAST::getFirstChild);
+        boolean hasWithAnnotation = false;
+        if (firstAnnotationChild.isPresent()) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java]: we do not need to execute this IF block if `hasBefore` is true. Please refactor the code.

Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java]: done

**Comment by romani** [comment]
Please regenerate diff report, and confirm that it is the same, no by to publish it, just confirm, I pretty sure it will be the same, but we had examples in our history when minor change result in regression.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java [hunk 4]
```diff
@@ -107,24 +107,41 @@ public void visitToken(DetailAST ast) {
...
+        return hasBefore || hasWithAnnotation(ast);
+    }
+
+    /**
+     * Checks javadoc existence in annotations list.
+     *
+     * @param ast package def
+     * @return true if there is a javadoc, false otherwise
+     */
+    private static boolean hasWithAnnotation(DetailAST ast) {
```
[the remark this answers is not in the document]
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java]: done

**Comment by strkkk** [comment]
@romani ping

**Comment by romani** [comment]
> Please regenerate diff report, and confirm that it is the same,

 not done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java [hunk 5]
```diff
@@ -107,24 +107,41 @@ public void visitToken(DetailAST ast) {
      * @return true if there is javadoc, false otherwise
      */
     private static boolean hasJavadoc(DetailAST ast) {
-        final boolean hasBefore = isJavadocComment(ast.getPreviousSibling());
-        // need to go 3 levels down the tree for annotation case
-        final boolean hasWithAnnotation = Optional.of(ast.getFirstChild())
-            .map(DetailAST::getFirstChild)
+        final boolean hasBefore = ast.getPreviousSibling() != null
+            && isJavadoc(ast.getPreviousSibling());
+        return hasBefore || hasJavadocAboveAnnotation(ast);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java]: please name it as `hasJavadocBefore`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java [hunk 6]
```diff
@@ -107,24 +107,41 @@ public void visitToken(DetailAST ast) {
...
+                    break;
+                }
+            }
+        }
+        return result;
     }
 
     /**
      * Checks that ast is a javadoc comment.
+     *
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java]: I do not mind ... but please move to separate commit.
Looks like just another idea for new Check.  Is this update from your IDE ?

Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java]: I do not understand what you mean

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java]: Please do other items.

This is just a side note that, non functional change is mixed with functional - not good.

**Comment by romani** [comment]
@strkkk , ping.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java (hunk 6 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java]: In some PRs I was asked to add such line to increase javadoc readability. I removed it

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java (hunk 5 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java]: done

**Comment by strkkk** [comment]
@romani I corrected items above and added link to updated report in first post

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java (hunk 6 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java]: yes, and it is completely ok to such clean up .... but there is nuance ... we do not mix non-fuctional refactoring with functional fixes in same commit, they should be in different commits (can be send in same PR).
split a single commit into two is easy:
```
git reset HEAD~1
git add -p 
# follow interactive mode by simply anwering y to non functional change
git commit -m "minor: code cleanup ...."
git add . 
git commit -m "Issue ......"
....
```

Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java]: I know how to split commits :) There is just no clear policy/check is empty line is required in javadoc

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java]: yes, we have no Check on this, so in code we have whatever. I think separation empty line is good.

**Comment by strkkk** [comment]
@romani
> you should have permission to do this

I don't.

Also, I do not have a permissions to mark conversations as resolved as a reviewer

**Comment by romani** [comment]
I granted you "Triage" permission, please recheck now.

> Also, I do not have a permissions to mark conversations as resolved as a reviewer

please do not resolve conversations, author of comment should resolve conversation as acknowledgement that item is actually done (after rechecking).

**Comment by strkkk** [comment]
> I granted you "Triage" permission, please recheck now.

Yeah, it is available now.

> author of comment should resolve conversation as acknowledgement that item is actually done (after rechecking)

This is exactly what I said :) if I was asked or started a review (so I added a comment and want to mark it as resolved after changes applied), I can't do that at the moment
